### ISSUE TITLE
Question:  how should we use parameter "fixed_tag"

### ISSUE THREAD

**Issue by gintuoklis** [HodorNV/ALOps#49]
Hey Guys, 

Could you please elaborate more how should we use the parameter "fixed_tag"? According to the documentation it "Allows recycling of docker containers." If I understand it right, it should not delete container but reuse it next time. This way build time should be decrease. Is it correct?

I tried to set it for all the build steps in my yml file, but did not see any improvement. My build pipleline is almost identical to 01_DefaultBuild.yml. (just few minor changes)

Best Regards, 
Gintautas

**Comment by AdminHodor** [comment]
Dear Gintautas,

When not using the "fixed_tag" containers are bound against a certain build or release and cannot be picked-up later from another pipeline run. The "fixed_tag" overwrites this default behavior and will force the job to pick the container which was created with the same "fixed_tag" rather than taking one bound to the build. 

>If I understand it right, it should not delete container but reuse it next time.
Not exactly, the "fixed_tag" only changes the way containers are bound to the pipeline, they do not influence the workflow. If you want to reuse the same container over and over again, you should exclude the "ALOpsDockerRemove" step from your pipeline. 

The parameter is intended/mostly used in release flows, to target fixed container which are setup as "QA, "STAGING", "UAT",....

You could use it in build pipelines to decrease build-time, but please note this could have undesirable effects since the container is not "fresh" and will require some pipeline changes to mitigate side effects. For example: if you use the pipeline to validate PullRequests you will quickly run into problems when using "fixed_tag".  PR1 for example added a field to a table, it builds and deploys the extension inside the docker container. Right after this, PR2 is triggered, with just a simple code modification (and not containing the field from PR1). The build for PR2 will run just fine right up to the publish step. The publish will fail because of breaking changing (trying to remove the field from PR1). You can mitigate this by adding the "ALOpsAppClean" step, again adding time to the build.

Long story short, "fixed_tag" is not meant to decrease pipeline time but rather target a know container. However it could be used for this purpose when accounting for possible side-effects. 

What are the times you are getting on "ALOpsDockerStart" and "ALOpsDockerWait" ? 
The time are directly bound to the performance of the agent, faster build agents do a faster job, which in our experience can make a big difference. (we ran +60.000 builds for performance measuring)
For example, for "ALOpsDockerStart" / "ALOpsDockerWait"
- Azure VM =>  15 sec / 125 sec
- OnPrem VM => 10 sec / 72 sec
- ALOps Agent => 5 sec / 36 sec

Kind regards,

Bart Bourgeois

**Comment by gintuoklis** [comment]
Hey @AdminHodor, 
Thanks for the detailed answer. 
Mine pipeline runs 5-6minutes on azure VM (Standard F4s (4 vcpus, 8 GiB memory))
With fixed_tag managed to get <2minutes.

Next question :) What is ALOps Agent?

**Comment by AdminHodor** [comment]
Hello @gintuoklis ,

ALOps Agent is a side project we are working on, in short we are aiming at "DevOps BC Agent As A Service". Purpose build hardware for getting the fastest pipeline possible. The agents have most current images pre-loaded. We started the project for internal use, for ALOps CI/CD we currently run about 1500 builds a month, so every minute counts. Currently we have it in a beta stage where we are focusing on security and scaling. Depending on the demand for it we will put in the effort to get this service to market with very competitive pricing. Our first priority is to get full coverage of partner scenario's with ALOps, ideally we aim for Q1 2020 to get the agent-as-a-service kickstarted. 

Kind regards,

Bart Bourgeois
